core/src/main/java/com/google/instrumentation/common/Timestamp.java

```diff
@@ -13,18 +13,24 @@
 
 package com.google.instrumentation.common;
 
+import com.google.common.math.LongMath;
+import java.math.RoundingMode;
+import javax.annotation.concurrent.Immutable;
+
 /**
  * A representation of an instant in time. The instant is the number of nanoseconds after the number
  * of seconds since the Unix Epoch.
  *
  * <p>Use {@link TimestampFactory#now} to get the current timestamp since epoch
  * (1970-01-01T00:00:00Z).
  */
+@Immutable
 public final class Timestamp {
   private static final long MAX_SECONDS = 315576000000L;
   private static final int MAX_NANOS = 999999999;
   private static final long NUM_MILLIS_PER_SECOND = 1000L;
   private static final int NUM_NANOS_PER_MILLI = 1000 * 1000;
+  private static final long NUM_NANOS_PER_SECOND = NUM_MILLIS_PER_SECOND * NUM_NANOS_PER_MILLI;
```

Comment:
Nit: I think the calculation is easier to parse/check if it's reversed, i.e.

private static final long NUM_NANOS_PER_SECOND = NUM_NANOS_PER_MILLI * NUM_MILLIS_PER_SECOND;